Nodejs/Product/Nodejs/SharedProject/CommonProjectNode.DiskMerger.cs

```diff
@@ -12,128 +15,226 @@ internal abstract partial class CommonProjectNode
-                            if (IsRecursiveSymLink(dir.Name, curDir))
+                            foreach (var curDir in Directory.EnumerateDirectories(dir.Path))
                             {
-                                // don't add recursive sym links
-                                continue;
+                                if (this.project.IsFileHidden(curDir))
+                                {
+                                    continue;
+                                }
+                                if (IsFileSymLink(curDir))
+                                {
+                                    if (IsRecursiveSymLink(dir.Path, curDir))
+                                    {
+                                        // don't add recursive sym links
+                                        continue;
+                                    }
+
+                                    // track symlinks, we won't get events on the directory
+                                    this.project.CreateSymLinkWatcher(curDir);
+                                }
+
+                                var folderNode = this.project.FindNodeByFullPath(CommonUtils.EnsureEndSeparator(curDir));
+                                if (folderNode == null)
+                                {
+                                    folderNode = this.InvokeOnUIThread(() => this.project.AddAllFilesFolder(dir.Parent, CommonUtils.EnsureEndSeparator(curDir), hierarchyCreated));
+                                }
+                                else
+                                {
+                                    missingChildren.Remove(folderNode);
+                                }
+                                this.dirs.Add(new DirState(curDir, folderNode));
                             }
+                        }
+                        catch (Exception exc) when (IsExpectedException(exc))
+                        {
+                            // directory was deleted, we don't have access, etc...
+                        }
 
-                            // track symlinks, we won't get events on the directory
-                            this.project.CreateSymLinkWatcher(curDir);
+                        var wasExpanded = this.project.ParentHierarchy != null ? dir.Parent.GetIsExpanded() : false;
+                        try
+                        {
+                            foreach (var file in Directory.EnumerateFiles(dir.Path))
+                            {
+                                if (this.project.IsFileHidden(file))
+                                {
+                                    continue;
+                                }
+                                var fileNode = this.project.FindNodeByFullPath(file);
+                                if (fileNode == null)
+                                {
+                                    fileNode = this.InvokeOnUIThread(() => this.project.AddAllFilesFile(dir.Parent, file));
```

Comment:
Is it worth trying to batch these? Seems like if a large folder (lots of files) is added, this is going to thrash back and forth between the UI & background thread to add them all individually.